src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepDynamicContext.java

```diff
@@ -0,0 +1,246 @@
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.support.steps;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Util;
+import hudson.init.InitMilestone;
+import hudson.model.Computer;
+import hudson.model.Executor;
+import hudson.model.Node;
+import hudson.model.Queue;
+import hudson.model.Result;
+import hudson.model.TaskListener;
+import hudson.remoting.VirtualChannel;
+import hudson.slaves.OfflineCause;
+import hudson.slaves.WorkspaceList;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.FilePathUtils;
+import org.jenkinsci.plugins.workflow.steps.DynamicContext;
+import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.support.DefaultStepContext;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Persistent representation for context of {@link ExecutorStepExecution}.
+ * Supersedes {@link FilePathDynamicContext} (never mind {@link org.jenkinsci.plugins.workflow.support.pickles.FilePathPickle}),
+ * {@link org.jenkinsci.plugins.workflow.support.pickles.ExecutorPickle},
+ * {@link org.jenkinsci.plugins.workflow.support.pickles.ComputerPickle},
+ * and {@link org.jenkinsci.plugins.workflow.support.pickles.WorkspaceListLeasePickle}.
+ */
+@Restricted(NoExternalUse.class)
+public final class ExecutorStepDynamicContext implements Serializable {
+
+    private static final Logger LOGGER = Logger.getLogger(ExecutorStepDynamicContext.class.getName());
+
+    private static final long serialVersionUID = 1;
+
+    final @NonNull ExecutorStepExecution.PlaceholderTask task;
+    final @NonNull String node;
+    private final @NonNull String path;
+    /** Non-null after {@link #resume} if all goes well. */
+    private transient @Nullable Executor executor;
+    /** Non-null after {@link #resume} if all goes well. */
+    private transient @Nullable WorkspaceList.Lease lease;
+
+    ExecutorStepDynamicContext(ExecutorStepExecution.PlaceholderTask task, WorkspaceList.Lease lease, Executor executor) {
+        this.task = task;
+        this.node = FilePathUtils.getNodeName(lease.path);
+        this.path = lease.path.getRemote();
+        this.executor = executor;
+        this.lease = lease;
+    }
+
+    void resume(StepContext context) throws Exception {
+        if (executor != null) {
+            throw new IllegalStateException("Already resumed");
+        }
+        while (Jenkins.get().getInitLevel() != InitMilestone.COMPLETED) {
```

Comment:
when is this called? if Jenkins fails to start will this leave a non deamon thread running that causes the Jenkins termination to fail?

Should this have some form of an exit path for `Jenkins.get().isTerminating()`?

Reply:
> a non deamon thread running that causes the Jenkins termination to fail

Not sure offhand if that is actually true, but I guess it should not hurt to check for the terminating flag. Will try it.